### PR TITLE
Fix SQL queries for inline datasource with null values

### PR DESCRIPTION
**Pull request by abhishekagarwal87**
Fixes a bug because of which some SQL queries cannot be parsed using druid convention. Specifically, these queries translate to an inline datasource and have some null values. Calcite internally uses NULL as SQL type for these literals and that is not supported by the druid. ~In this PR, we are marking such null literals as the STRING column type.~

Update - Instead of checking for null type, I am now allowing null column types to be returned while building RowSignature in `org.apache.druid.sql.calcite.table.RowSignatures#fromRelDataType`. RowSignature already allows null column type for any column. Doing so should also fix bindable queries such as `select (1,2)`. When such queries are run with `headers` set to true, we get an exception in `org.apache.druid.sql.http.ArrayWriter#writeHeader`. This is again a similar exception to the one addressed in this PR. Because SQL type for the result column is `RECORD` and that doesn't have a corresponding columnType. 

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java [hunk 1]
```diff
@@ -128,6 +128,11 @@ static Object getValueFromLiteral(RexLiteral literal, PlannerContext plannerCont
       case TIMESTAMP:
       case DATE:
         return Calcites.calciteDateTimeLiteralToJoda(literal, plannerContext.getTimeZone()).getMillis();
+      case NULL:
```
Comment (LakshSingla) [inline sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java:131]: Could this have been the case when Boolean is of null type? 
Referring to this conversation: https://github.com/apache/druid/pull/11923#discussion_r766485610

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/Expressions.java [hunk 2]
```diff
@@ -242,6 +242,7 @@ private static DruidExpression rexCallToDruidExpression(
                                                            .lookupOperatorConversion(operator);
 
     if (conversion == null) {
+      plannerContext.setPlanningError("SQL query requires '%s' operator that is not supported.", operator.getName());
```
Comment (LakshSingla) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/Expressions.java:245]: Since this method is mostly used as a helper/util method, setting the error here might not be desirable in every use case. Would it be better if it is moved to a place like this?  [Projections, line 105](https://github.com/apache/druid/blob/476d0bf4be4199e97695bd568d165cda98523d37/sql/src/main/java/org/apache/druid/sql/calcite/rel/Projection.java#L105)

Comment (abhishekagarwal87) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/Expressions.java:245]: the absence of a corresponding operator seems like a genuine feature gap. Also, I can populate a more precise error here than I can in the uppar part of the call stack. Since a null can be returned for other reasons.

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java (hunk 1 above)
Comment (abhishekagarwal87) [reply sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java:131]: If it was, wouldn't we see an unsupported type exception? do you mean something different

**Comment by abhishekagarwal87** [comment]
@LakshSingla thank you for your review. Please let me know if you feel there is a code change required.

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java (hunk 1 above)
Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java:131]: It seems I mixed up `literal.getType().getSqlTypeName()` and `literal.getSqlTypeName()` when I originally wanted to ask the question. 

In the current modification, instead of returning `null`, should it be `throw new UnsupportedSQLQueryException("%s type is not supported", literal.getType().getSqlTypeName());`, to preserve the original behavior.

**Comment by LakshSingla** [comment]
Minor doubt, thanks for addressing the comments. LGTM.

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java (hunk 1 above)
Comment (abhishekagarwal87) [reply sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java:131]: that's what it was doing earlier which means we can't support null literals in SQL queries.

Comment (abhishekagarwal87) [reply sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java:131]: this change fixes that exception.

Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java:131]: Got it, thanks. I was incorrectly assuming that this was aimed at making the error message more specific to null columns.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java [hunk 3]
```diff
@@ -173,6 +173,8 @@ public static ColumnType getValueTypeForRelDataTypeFull(final RelDataType type)
         return ColumnType.LONG_ARRAY;
       }
       return ColumnType.STRING_ARRAY;
+    } else if (sqlTypeName == SqlTypeName.NULL) {
+      return ColumnType.STRING;
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:177]: why is this needed?

Comment (abhishekagarwal87) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:177]:
```
org.apache.druid.java.util.common.ISE: Cannot translate sqlTypeName[NULL] to Druid type for field[EXPR$0]
	at org.apache.druid.sql.calcite.table.RowSignatures.fromRelDataType(RowSignatures.java:68)
	at org.apache.druid.sql.calcite.rule.DruidLogicalValuesRule.onMatch(DruidLogicalValuesRule.java:75)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:208)
```
We want to build the DruidTable object with a valid RowSignature from the inline datasource. RowSignature needs a non-null column type for each column. When there is just row with null value in the inline data, the sql type is null. When we don't have a corresponding druid column type, we throw the above exception. 

Instead of introducing a new column type, I have mapped the null type to string type.

Comment (abhishekagarwal87) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:177]: Correction - RowSignature allows non-null column type. The exception is thrown during the conversion of signature from calcite type to druid type.

**Review by clintropolis — APPROVED** [review]
